cyto_dl/nn/vits/blocks/masked_unit_attention.py

```diff
@@ -0,0 +1,232 @@
+        attn_drop : float, optional
+            Dropout rate for attention, by default 0.0.
+        proj_drop : float, optional
+            Dropout rate for projection, by default 0.0.
+        q_stride : List[int], optional
+            Stride for query, by default [1, 1, 1].
+        patches_per_mask_unit : List[int], optional
+            Number of patches per mask unit, by default [2, 12, 12].
+        """
+        super().__init__()
+        patches_per_mask_unit, q_stride = validate_spatial_dims(
+            spatial_dims, [patches_per_mask_unit, q_stride]
+        )
+
+        self.spatial_dims = spatial_dims
+        self.dim = dim
+        self.dim_out = dim_out
+        self.q_stride = q_stride
+
+        self.norm1 = norm_layer(dim)
+
+        do_pool = np.any(np.array(q_stride) > 1) or dim != dim_out
+
+        self.attn = MaskUnitAttention(
+            dim,
+            dim_out,
+            spatial_dims=spatial_dims,
+            num_heads=heads,
+            q_stride=q_stride,
+            patches_per_mask_unit=patches_per_mask_unit,
+        )
+
+        self.norm2 = norm_layer(dim_out)
+        self.mlp = Mlp(dim_out, int(dim_out * mlp_ratio), act_layer=act_layer)
+
+        self.drop_path = DropPath(drop_path) if drop_path > 0 else nn.Identity()
+
+        # max pooling by q stride within a mask unit
+        if self.spatial_dims == 3:
+            skip_connection_pooling = Reduce(
+                "b n (n_patches_z q_stride_z n_patches_y q_stride_y n_patches_x q_stride_x) c -> b n (n_patches_z n_patches_y n_patches_x) c",
+                reduction="mean",
+                q_stride_z=self.q_stride[0],
+                q_stride_y=self.q_stride[1],
+                q_stride_x=self.q_stride[2],
+                n_patches_z=self.attn.pooled_patches_per_mask_unit[0],
+                n_patches_y=self.attn.pooled_patches_per_mask_unit[1],
+                n_patches_x=self.attn.pooled_patches_per_mask_unit[2],
+            )
+        elif self.spatial_dims == 2:
+            skip_connection_pooling = Reduce(
+                "b n (n_patches_y q_stride_y n_patches_x q_stride_x) c -> b n (n_patches_y n_patches_x) c",
+                reduction="mean",
+                q_stride_y=self.q_stride[0],
+                q_stride_x=self.q_stride[1],
+                n_patches_y=self.attn.pooled_patches_per_mask_unit[0],
+                n_patches_x=self.attn.pooled_patches_per_mask_unit[1],
+            )
+
+        self.proj = (
+            torch.nn.Sequential(skip_connection_pooling, nn.Linear(dim, dim_out))
+            if do_pool
+            else torch.nn.Identity()
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        x: batch x mask units x tokens x emb_dim
+        """
+        # Attention + Q Pooling
+        x_norm = self.norm1(x)
+        # change dimension and subsample within mask unit for skip connection
+        x = self.proj(x_norm)
+
+        x = x + self.drop_path(self.attn(x_norm))
```

Comment:
Does dim_out = dim for skip connection with attention?

Reply:
Good question - each block specified in the architecture argument doubles the embedding dimension and halves the size of the mask unit. This doubling/pooling happens on the last repeat of the block, so dim_out=dim for all repeats except the last. I updated the docstring with an example.

Reply:
cool!